### DISCUSSION TITLE
Hive can't update transactional table which created by trino

### DISCUSSION THREAD

**Discussion by boson234** [trinodb/trino#16488] [Q&A] [answered]
hi, 
    I create a trancational table with trino 410 ,and insert some rows, and I can update rows with trion success.
howerver, when I try to update data for this table with hive,it failed.
**_my hive version is 3.1.0_**
create table trino_11 with (transactional =true) as select * from (values(1),(2),(3)) as t1(a);
![image](https://user-images.githubusercontent.com/34983864/224291449-209ba622-ff62-4acb-9a50-218aa9cc167c.png)
![image](https://user-images.githubusercontent.com/34983864/224291792-a9ddfb9c-b554-4aa8-95b5-e46da7e086e0.png)

![image](https://user-images.githubusercontent.com/34983864/224291667-efd90036-e3fd-4c50-95ae-4580046345e8.png)

**_here is the error log on Mapreduce_**
![image](https://user-images.githubusercontent.com/34983864/224291360-55962a98-6185-4991-a27f-755fbcbea6f2.png)

**Comment by bitsondatadev** [comment]
Have you created an equivalent table in Hive and it was successful?

**Comment by boson234** [comment]
yes，all that in hive works.



---Original---
From: ***@***.***&gt;
Date: Fri, Mar 10, 2023 19:38 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [trinodb/trino] Hive can't update transactional table whichcreated by trino (Discussion #16488)




 
Have you created an equivalent table in Hive and it was successful?

**Comment by bitsondatadev** [comment]
@djsstarburst or @findepi, any ideas as to what could be causing this? I believe this worked on a previous version of Hive.

**Comment by boson234** [comment, marked as answer]
bucket count!!!  **trino is 0 but hive is -1**
![image](https://user-images.githubusercontent.com/34983864/224470732-80148b30-b4bc-466b-9a1b-957d10eda70f.png)
![image](https://user-images.githubusercontent.com/34983864/224470792-50fd5830-cf7c-4af0-9735-23e429c41978.png)

I add codes to set NumBucket to -1 
in ThriftMetastoreUtil.makeStorageDescriptor
and it works

**Comment by djsagain** [comment]
I was unable to reproduce the UPDATE failure.  This test added to `TestHiveTransactionalTable` passes.  I'm running Trino tip master and Hive version 3.1.2-20:

```
    @Test
    public void testIncorrectBucketCount()
    {
        String tableName = "trino_11_" + randomNameSuffix();
        onTrino().executeQuery("create table %s with (transactional = true) as select * from (values(1),(2),(3)) as t1(a)".formatted(tableName));
        onHive().executeQuery("UPDATE %s SET a = a + 4 WHERE a = 1".formatted(tableName));
        verifySelectForTrinoAndHive("SELECT * FROM " + tableName, row(5), row(2), row(3));
        onTrino().executeQuery("UPDATE %s SET a = a + 4 WHERE a = 5".formatted(tableName));
        verifySelectForTrinoAndHive("SELECT * FROM " + tableName, row(9), row(2), row(3));
        onTrino().executeQuery("DROP TABLE " + tableName);
    }
```

**Reply by boson234** [reply]
> I was unable to reproduce the UPDATE failure. This test added to `TestHiveTransactionalTable` passes. I'm running Trino tip master and Hive version 3.1.2-20:
> 
> ```
>     @Test
>     public void testIncorrectBucketCount()
>     {
>         String tableName = "trino_11_" + randomNameSuffix();
>         onTrino().executeQuery("create table %s with (transactional = true) as select * from (values(1),(2),(3)) as t1(a)".formatted(tableName));
>         onHive().executeQuery("UPDATE %s SET a = a + 4 WHERE a = 1".formatted(tableName));
>         verifySelectForTrinoAndHive("SELECT * FROM " + tableName, row(5), row(2), row(3));
>         onTrino().executeQuery("UPDATE %s SET a = a + 4 WHERE a = 5".formatted(tableName));
>         verifySelectForTrinoAndHive("SELECT * FROM " + tableName, row(9), row(2), row(3));
>         onTrino().executeQuery("DROP TABLE " + tableName);
>     }
> ```

maybe the hive version I used has some difference from open source hive.
